### PR TITLE
[YUNIKORN-1107] Make health check occur in the background

### PR DESCRIPTION
**Pull request by lowc1012**
### What is this PR for?
* Change "ws/v1/scheduler/healthcheck" API to fetch health check cache in scheduler context
* Run health check periodically in the background

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
[YUNIKORN-1107](https://issues.apache.org/jira/projects/YUNIKORN/issues/YUNIKORN-1107)

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/scheduler/health_checker.go [hunk 1]
```diff
@@ -20,14 +20,73 @@ package scheduler
...
+	}()
+}
+
+//nolint:unused
+func (c *HealthChecker) stop() {
+	c.stopChan <- struct{}{}
+	close(c.stopChan)
+}
+
+func UpdateSchedulerHealthStatusCache(latest dao.SchedulerHealthDAOInfo, schedulerContext *ClusterContext) {
```
Comment (pbacsko) [inline pkg/scheduler/health_checker.go:80]: Is this method neet to be exported? (capital "u")

File: pkg/scheduler/context.go [hunk 2]
```diff
@@ -918,3 +921,11 @@ func (cc *ClusterContext) SetRMInfo(rmID string, rmBuildInformation map[string]s
...
 	}
 }
+
+func (cc *ClusterContext) GetHealthCheckCache() *dao.SchedulerHealthDAOInfo {
+	return cc.healthCheckCache
+}
+
+func (cc *ClusterContext) SetHealthCheckCache(c *dao.SchedulerHealthDAOInfo) {
+	cc.healthCheckCache = c
+}
```
Comment (pbacsko) [inline pkg/scheduler/context.go:931]: These methods are invoked from different goroutines. Add the necessary `RLock()` and `RUnlock()` calls.

File: pkg/scheduler/context.go [hunk 3]
```diff
@@ -54,6 +55,8 @@ type ClusterContext struct {
 	startTime time.Time
 
 	sync.RWMutex
+
+	healthCheckCache *dao.SchedulerHealthDAOInfo
```
Comment (pbacsko) [inline pkg/scheduler/context.go:59]: The name "lastHealthCheckResult" sounds more appropriate to me. "cache" makes me think that it's something more complicated, like a data structure or something.

**Comment by lowc1012** [comment]
@pbacsko Thanks for your review!
The above problems have been solved.

File: pkg/webservice/handlers.go [hunk 4]
```diff
@@ -516,14 +516,23 @@ func isChecksumEqual(checksum string) bool {
...
+
+	// Fetch last healthCheck result
+	result := schedulerContext.GetLastHealthCheckResult()
+	if result != nil {
+		if !result.Healthy {
+			log.Logger().Error("Scheduler is not healthy", zap.Any("health check info", *result))
+			buildJSONErrorResponse(w, "Scheduler is not healthy", http.StatusServiceUnavailable)
+		} else {
+			log.Logger().Info("Scheduler is healthy", zap.Any("health check info", *result))
+			buildJSONErrorResponse(w, "Scheduler is healthy", http.StatusOK)
```
Comment (HuangTing-Yao) [inline pkg/webservice/handlers.go:528]: When scheduler is healthy, we shouldn't log it with `ERROR` level.

**Review by HuangTing-Yao — CHANGES_REQUESTED** [review]

File: pkg/webservice/handlers.go (hunk 4 above)
Comment (lowc1012) [reply pkg/webservice/handlers.go]: done

**Comment by HuangTing-Yao** [comment]
The codecov check doesn't triggered, also we can add `TestHealthCheckerStartStop()` for `start()` and `stop()` to make sure the panic won't happened.

**Review by HuangTing-Yao — APPROVED** [review]
LGTM

**Pull request merged**